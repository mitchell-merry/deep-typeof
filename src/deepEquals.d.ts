declare module "deep-equals" {
	function deepEquals (a: unknown, b: unknown): boolean;
	export = deepEquals;
}