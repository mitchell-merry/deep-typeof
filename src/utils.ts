// imp/ort deepEquals from "deep-equals";
const deepEquals = require("deep-equals"); // require to prevent "They are the same" text

export function unique<T>(target: T, index: number, array: T[]) {
	return !array.some((curr, currIndex) => currIndex < index && deepEquals(curr, target));
}