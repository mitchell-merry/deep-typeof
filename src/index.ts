import { unique } from "./utils";

export function deepTypeof(value: unknown): unknown {
	if (typeof value !== "object") {
		return typeof value;
	}

	if (value === null) {
		return null;
	}

	if (Array.isArray(value)) {
		return value.map(deepTypeof).filter(unique);
	}

	const objectKeysTypes: Record<string, unknown> = {};
	for(const key in value) {
		objectKeysTypes[key] = deepTypeof((value as any)[key]);
	}

	return objectKeysTypes;
}