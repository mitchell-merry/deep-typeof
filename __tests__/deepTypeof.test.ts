import { describe, it, expect } from "vitest";
import { deepTypeof } from "../src/index";

describe('deepTypeof', () => {
	it('Uses typeof for primitives', () => {
		const data = [ "hello", 4, BigInt("100"), undefined, () => {}, Symbol() ];
		expect(data.map(deepTypeof)).toEqual([ "string", "number", "bigint", "undefined", "function", "symbol" ]);
	});
	
	it('Types an object one level down', () => {
		const data = {
			"hiya": "myString",
			4: 545
		};
	
		expect(deepTypeof(data)).toEqual({
			"hiya": "string",
			"4": "number"
		});
	});
	
	
	it('Types an array', () => {
		const data = [
			"myString",
			234,
			() => {},
			"anotherString"
		];
	
		expect(deepTypeof(data)).toEqual([
			"string",
			"number",
			"function"
		]);
	});
	
	it('array of objects', () => {
		const data = [
			{ id: "mmerry" },
			{ id: "bmerry" },
		];
	
		expect(deepTypeof(data)).toEqual([
			{ id: "string" }
		]);
	});

	class MyClass {
		prop: string;
		constructor() { this.prop = "hi"; }
	}
	
	it('instance of a class', () => {
		expect(deepTypeof(new MyClass())).toEqual({
			prop: "string"
		});
	});

	class MySubclass extends MyClass {
		prop2: number;
		instanceMember: MyClass;
		
		constructor() {
			super();
			this.prop2 = 4;
			this.instanceMember = new MyClass();
		}

		async anFunc() { }
	}
	
	it('instance of a subclass with instances as members', () => {
		expect(deepTypeof(new MySubclass())).toEqual({
			prop: "string",
			prop2: "number",
			instanceMember: {
				prop: "string"
			}
		});
	});

	it('snapshot test1', () => {
		const data = require('./__snapshots__/test1.json');
		expect(deepTypeof(data)).toMatchSnapshot();
	})
});