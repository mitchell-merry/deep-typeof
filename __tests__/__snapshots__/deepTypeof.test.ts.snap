// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deepTypeof > snapshot test1 1`] = `
[
  {
    "_id": "string",
    "about": "string",
    "address": "string",
    "age": "number",
    "balance": "string",
    "company": "string",
    "email": "string",
    "eyeColor": "string",
    "favoriteFruit": "string",
    "friends": [
      {
        "id": "number",
        "name": "string",
      },
    ],
    "gender": "string",
    "greeting": "string",
    "guid": "string",
    "index": "number",
    "isActive": "boolean",
    "latitude": "number",
    "longitude": "number",
    "name": "string",
    "phone": "string",
    "picture": "string",
    "registered": "string",
    "tags": [
      "string",
    ],
  },
]
`;
